driver/src/main/com/mongodb/GridFSDownloadStreamImpl.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb;
+
+import com.mongodb.client.GridFSDownloadStream;
+import com.mongodb.client.MongoCollection;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonObjectId;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import static com.mongodb.assertions.Assertions.isTrue;
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+class GridFSDownloadStreamImpl extends InputStream implements GridFSDownloadStream {
+    private final BsonDocument fileInfo;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private final BsonObjectId fileId;
+    private final long lengthInBytes;
+    private final int chunkSizeInBytes;
+    private final int numberOfChunks;
+    private int chunkIndex;
+    private int bufferOffset;
+    private long currentPosition;
+    private byte[] buffer = null;
+    private boolean closed;
+
+    GridFSDownloadStreamImpl(final BsonDocument fileInfo, final MongoCollection<BsonDocument> chunksCollection) {
+        this.fileInfo = notNull("file information", fileInfo);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+
+        try {
+            isTrue("file information contains _id", fileInfo.containsKey("_id") && fileInfo.get("_id").isObjectId());
+            isTrue("file information contains length", fileInfo.containsKey("length") && fileInfo.get("length").isInt64());
+            isTrue("file information contains chunkSize", fileInfo.containsKey("chunkSize") && fileInfo.get("chunkSize").isInt32());
+        } catch (IllegalStateException e) {
+            throw new MongoGridFSException("GridFS file information is not in the expected format", e);
+        }
+        fileId = fileInfo.getObjectId("_id");
+        lengthInBytes = fileInfo.getInt64("length").getValue();
+        chunkSizeInBytes = fileInfo.getInt32("chunkSize").getValue();
+        numberOfChunks = (int) Math.ceil((double) lengthInBytes / chunkSizeInBytes);
+    }
+
+    @Override
+    public BsonDocument getFileInformation() {
```

Comment:
Usability would be improved if we had a GridFSFile class instead of BsonDocument.  I suppose the metadata would still have to be BsonDocument.